packages/drivers/odsp-driver/src/getFileLink.ts

```diff
@@ -90,6 +93,86 @@ export async function getFileLink(
 	return fileLink;
 }
 
+async function getRequestInformation(
+	getToken: TokenFetcher<OdspResourceTokenFetchOptions>,
+	odspUrlParts: IOdspUrlParts,
+	logger: ITelemetryLoggerExt,
+	options: TokenFetchOptionsEx,
+	fileItem: FileItemLite,
+) {
+	const storageTokenFetcher = toInstrumentedOdspTokenFetcher(
+		logger,
+		odspUrlParts,
+		getToken,
+		true /* throwOnNullToken */,
+	);
+	const storageToken = await storageTokenFetcher(options, "GetFileLinkCore");
+	assert(
+		storageToken !== null,
+		0x2bb /* "Instrumented token fetcher with throwOnNullToken = true should never return null" */,
+	);
+
+	// IMPORTANT: In past we were using GetFileByUrl() API to get to the list item that was corresponding
+	// to the file. This was intentionally replaced with GetFileById() to solve the following issue:
+	// GetFileByUrl() uses webDavUrl to locate list item. This API does not work for Consumer scenarios
+	// where webDavUrl is constructed using legacy ODC format for backward compatibility reasons.
+	// GetFileByUrl() does not understand that format and thus fails. GetFileById() relies on file item
+	// unique guid (sharepointIds.listItemUniqueId) and it works uniformly across Consumer and Commercial.
+	const { url, headers } = getUrlAndHeadersWithAuth(
+		`${
+			odspUrlParts.siteUrl
+		}/_api/web/GetFileById(@a1)/ListItemAllFields/GetSharingInformation?@a1=guid${encodeURIComponent(
+			`'${fileItem.sharepointIds.listItemUniqueId}'`,
+		)}`,
+		storageToken,
+		true,
+	);
+	const requestInit = {
+		method: "POST",
+		headers: {
+			"Content-Type": "application/json;odata=verbose",
+			"Accept": "application/json;odata=verbose",
+			...headers,
+		},
+	};
+
+	return { url, requestInit };
+}
+
+async function fetchWithLocationRedirectionHandling<T>(
+	getToken: TokenFetcher<OdspResourceTokenFetchOptions>,
+	odspUrlParts: IOdspUrlParts,
+	logger: ITelemetryLoggerExt,
+	options: TokenFetchOptionsEx,
+	fileItem: FileItemLite,
+): Promise<IOdspResponse<Response>> {
+	for (;;) {
+		const { url, requestInit } = await getRequestInformation(
+			getToken,
+			odspUrlParts,
+			logger,
+			options,
+			fileItem,
+		);
+
+		try {
+			return await fetchHelper(url, requestInit);
+		} catch (error: any) {
+			if (
+				error.errorType !== DriverErrorTypes.fileNotFoundOrAccessDeniedError ||
```

Comment:
Checking for `fileNotFoundOrAccessDeniedError` error type seems unrelated to this change.

Reply:
That is actually the error type we get when throwing a OdspRedirectError that I myself throw on odspErrorUtils.ts. 
![image](https://github.com/microsoft/FluidFramework/assets/10786508/642c771d-54cc-409a-94cf-062bcdca7c99)

I'm checking for both that error that I throw, as well as that there is a redirect URL available before doing any redirect logic

Reply:
Consider doing the following:

```ts

class OdspRedirectError {
	public static errorType = DriverErrorTypes.fileNotFoundOrAccessDeniedError;
	readonly errorType = OdspRedirectError.type;
...
}

if (error?.errorType !== OdspRedirectError .Type || !error?.redirectLocation)  {
    throw error;
}

```

or even better - way more readable.

class OdspRedirectError {
	public static errorType = DriverErrorTypes.fileNotFoundOrAccessDeniedError;
	readonly errorType = OdspRedirectError.type;

        static public isRedirectError(error: unknown) {
             return error?.errorType  === OdspRedirectError.errorType && !!error?.redirect;
        }
...
}

if (!OdspRedirectError(error)) {
    throw error;
}
...
}

Notice that error does not have to be object (i.e. it can be undefined, null, etc. - we never know)